DeepDiveLogCompiler.scala

```diff
@@ -170,29 +177,68 @@ class CompilationState( statements : DeepDiveLog.Program )  {
         ${ whereClauseStr }"""
   }
 
+  // Analyze the block visibility among statements 
+  def analyzeVisable(statements: List[Statement]) = {
+    val extractionRules   = new ListBuffer[ExtractionRule]()
+    val functionCallRules = new ListBuffer[FunctionCallRule]()
+    val inferenceRules    = new ListBuffer[InferenceRule]()
+    
+    statements foreach (_ match {
+      case s: ExtractionRule   => extractionRules += s
+      case s: FunctionCallRule => functionCallRules += s
+      case s: InferenceRule    => inferenceRules += s
+      case _                   => 
+    })
+
+    val extractionRulesGroup   = extractionRules.groupBy(_.q.head.name)
+    val functionCallRulesGroup = functionCallRules.groupBy(_.input)
+    val inferenceRulesGroup    = inferenceRules.groupBy(_.q.head.name)
+
+    extractionRulesGroup   foreach {keyVal => visable += keyVal._2(0)}
+    functionCallRulesGroup foreach {keyVal => visable += keyVal._2(0)}
+    inferenceRulesGroup    foreach {keyVal => visable += keyVal._2(0)}
+  }
```

Comment:
I don't understand why we repeat blocks of similar code three times (in analyzeVisable, analyzeDependency, and run) that basically constructs a map from name to a set of rules.  Instead, we can first construct a map only once (in a new function, e.g., named groupByHead), then use that map's values to populate the dependency (from analyzeDependency), and compile each key to produce the blocks (from run).  The map doesn't have to be of type `Map[String,List[Statement]]`.  Because we eventually want a representative statement from each group, the final map can be of type `Map[Statement,List[Statement]]` where the key is for example the first one in each group.  Can you clean up the redundant code please?  Is the description clear enough?